File: scripts/raku/dependancy_update.pl
use strict;
use warnings;
use TOML::Tiny qw(from_toml to_toml);

sub get_latest_version {
    my ($library) = @_;
    my $output = qx(cargo search $library);
    my ($new_ver) = $output =~ /"([^"]+)"/; # Capture version in quotes
    return $new_ver;
}

sub update_library_versions {
    my ($filename) = @_;
    open my $fh, '<', $filename or die "Cannot open $filename: $!";
    my $file_content = do { local $/; <$fh> };
    close $fh;

    my $toml = from_toml($file_content);

    if (exists $toml->{dependencies}) {
        while (my ($lib, $version) = each %{$toml->{dependencies}}) {
            my $new_version = get_latest_version($lib);
            if (defined $new_version and ($version ne $new_version)) {
                print "Updating $lib: $version -> $new_version\n";
                $toml->{dependencies}{$lib} = $new_version;
            } else {
                print "Latest version for $lib not found or already up to date.\n";
            }
        }

        my $new_toml_content = to_toml($toml);
        open my $fh_out, '>', $filename or die "Cannot open $filename for writing: $!";
        print $fh_out $new_toml_content;
        close $fh_out;
    } else {
        print "'dependencies' section not found in $filename\n";
    }
}

sub main {
    my ($filename) = @ARGV;
    update_library_versions($filename);
}

main() unless caller;

File: scripts/raku/dependancy_update.py
import toml
import subprocess
import re
import sys

def get_latest_version(library):
    """Fetch the latest version of a library using cargo search."""
    try:
        output = subprocess.check_output(['cargo', 'search', library], text=True)
        # Extract the version from the cargo search output
        match = re.search(r'"([^"]+)"', output)
        if match:
            return match.group(1)
    except subprocess.CalledProcessError as e:
        print(f"Error fetching version for {library}: {e}")
    return None

def update_library_versions(filename):
    """Update library versions in the TOML file."""
    with open(filename, 'r') as file:
        data = toml.load(file)

    dependencies = data.get('dependencies', {})
    for lib, version in dependencies.items():
        # Handle both plain version strings and complex structures with version fields
        if isinstance(version, str):
            old_version = version
        elif isinstance(version, dict) and 'version' in version:
            old_version = version['version']
        else:
            print(f"Skipping {lib}: unsupported format")
            continue

        new_version = get_latest_version(lib)
        if new_version and new_version != old_version:
            print(f"Updating {lib}: {old_version} -> {new_version}")
            if isinstance(version, str):
                dependencies[lib] = new_version
            elif isinstance(version, dict):
                version['version'] = new_version
        else:
            print(f"No update required or new version not found for {lib}.")

    with open(filename, 'w') as file:
        toml.dump(data, file)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python script.py <path/to/TOMLfile.toml>")
        sys.exit(1)
    filename = sys.argv[1]
    update_library_versions(filename)
